src/qibojit/backends/clifford_operations_gpu.py

```diff
@@ -0,0 +1,726 @@
+extern "C"
+__global__ void apply_rowsum(bool* symplectic_matrix, const long* h, const long* i, const int nqubits, const bool determined, const int nrows, long* g_exp, const int dim) {{
+    _apply_rowsum(symplectic_matrix, h, i, nqubits, determined, nrows, g_exp, dim);
+}}
+"""
+
+apply_rowsum = cp.RawKernel(apply_rowsum, "apply_rowsum", options=("--std=c++11",))
+
+
+def _rowsum(symplectic_matrix, h, i, nqubits, determined=False):
+    dim = _get_dim(nqubits)
+    nrows = len(h)
+    exp = cp.zeros(len(h), dtype=int)
+    apply_rowsum(
+        GRIDDIM_2D,
+        (BLOCKDIM,),
+        (symplectic_matrix, h, i, nqubits, determined, nrows, exp, dim),
+    )
+    return symplectic_matrix
+
+
+def _get_p(state, q, nqubits):
+    dim = _get_dim(nqubits)
+    return state.reshape(dim, dim)[nqubits:-1, q].nonzero()[0]
+
+
+def _random_outcome(state, p, q, nqubits):
+    dim = _get_dim(nqubits)
+    p = p[0] + nqubits
+    idx_pq = p * dim + q
+    tmp = state[idx_pq].copy()
+    state[idx_pq] = False
+    h = state.reshape(dim, dim)[:-1, q].nonzero()[0]
+    state[idx_pq] = tmp
+    if h.shape[0] > 0:
+        state = _rowsum(
+            state,
+            h,
+            p.astype(cp.uint) * cp.ones(h.shape[0], dtype=np.uint),
+            nqubits,
+            False,
+        )
+    state = state.reshape(dim, dim)
+    state[p - nqubits, :] = state[p, :]
+    outcome = cp.random.randint(2, size=None, dtype=cp.uint)
+    state[p, :] = False
+    state[p, -1] = outcome.astype(bool)
+    state[p, nqubits + q] = True
+    return state.ravel(), outcome
+
+
+def _determined_outcome(state, q, nqubits):
+    dim = _get_dim(nqubits)
+    state = state.reshape(dim, dim)
+    state[-1, :] = False
+    idx = state[:nqubits, q].nonzero()[0] + nqubits
+    state = _rowsum(
+        state.ravel(),
+        (2 * nqubits * cp.ones(idx.shape, dtype=np.uint)).astype(np.uint),
+        idx.astype(np.uint),
+        nqubits,
+        True,
+    )
+    return state, state[dim * dim - 1].astype(cp.uint)
+
+
+def cast(x, dtype=None, copy=False):
+    if dtype is None:
+        dtype = "complex128"
+    if cp.sparse.issparse(x):
+        if dtype != x.dtype:
+            return x.astype(dtype)
+        else:
+            return x
+    elif sparse.issparse(x):
+        cls = getattr(cp.sparse, x.__class__.__name__)
+        return cls(x, dtype=dtype)
+    elif isinstance(x, cp.ndarray) and copy:
+        return cp.copy(cp.asarray(x, dtype=dtype))
+    else:
+        return cp.asarray(x, dtype=dtype)
+
+
+def _clifford_pre_execution_reshape(state):
+    return state.ravel()
+
+
+def _clifford_post_execution_reshape(state, nqubits):
+    dim = _get_dim(nqubits)
+    return state.reshape(dim, dim)
+
+
+def identity_density_matrix(nqubits, normalize: bool = True):
+    n = 1 << nqubits
+    state = cp.eye(n, dtype="complex128")
+    cp.cuda.stream.get_current_stream().synchronize()
+    if normalize:
+        state /= 2**nqubits
+    return state.reshape((n, n))
```

Comment:
Here I'd say that using `cp.ravel` and `cp.reshape` is more robust just in case (I know, I know...)